src/visitors.rs

```diff
@@ -26,30 +28,139 @@ pub struct MirVisitor<'a, 'b: 'a, 'tcx: 'b> {
 impl<'a, 'b: 'a, 'tcx: 'b> MirVisitor<'a, 'b, 'tcx> {
-    /// Visits each basic block in order of occurrence.
+    pub fn new(crate_context: MirVisitorCrateContext<'a, 'b, 'tcx>) -> MirVisitor<'a, 'b, 'tcx> {
+        MirVisitor {
+            buffered_diagnostics: crate_context.buffered_diagnostics,
+            emit_diagnostic: crate_context.emit_diagnostic,
+            session: crate_context.session,
+            tcx: crate_context.tcx,
+            def_id: crate_context.def_id,
+            mir: crate_context.mir,
+            constant_value_cache: crate_context.constant_value_cache,
+            summary_cache: crate_context.summary_cache,
+
+            check_for_errors: false,
+            current_environment: Environment::default(),
+            current_span: syntax_pos::DUMMY_SP,
+            exit_environment: Environment::default(),
+            inferred_preconditions: List::new(),
+            path_conditions: List::new(),
+            post_conditions: List::new(),
+            preconditions: List::new(),
+            unwind_condition: List::new(),
+        }
+    }
+
+    /// Analyze the body and store a summary of its behavior in self.summary_cache.
     pub fn visit_body(&mut self) {
         debug!("visit_body({:?})", self.def_id);
+        // in_state[bb] is the join (or widening) of the out_state values of each predecessor of bb
+        let mut in_state: HashMap<rustc::mir::BasicBlock, Environment> = HashMap::new();
+        // out_state[bb] is the environment that results from analyzing block bb, given in_state[bb]
+        let mut out_state: HashMap<rustc::mir::BasicBlock, Environment> = HashMap::new();
+        // The entry block has no predecessors and its initial state is the function parameters.
+        let mut first_block = true;
+        for bb in self.mir.basic_blocks().indices() {
+            let i_state = if first_block {
+                first_block = false;
+                Environment::with_parameters(self.def_id, self.tcx)
+            } else {
+                Environment::default() // Will be joined with other states before analysis of bb
+            };
+            in_state.insert(bb, i_state);
+            out_state.insert(bb, Environment::default());
+        }
 
+        // Compute a fixed point, which is a value of out_state that will not grow with more iterations.
+        let mut changed = true;
+        let mut iteration_count = 0;
+        while changed {
+            changed = false;
+            for bb in self.mir.basic_blocks().indices() {
+                // Merge output states of predcessors of bb
+                let mut i_state = Environment::default();
+                for pred_bb in self.mir.predecessors_for(bb).iter() {
+                    // todo: obtain join condition from out_state[pred_bb]
+                    let join_condition = &abstract_value::TOP;
+                    if iteration_count < 3 {
```

Comment:
Heh, number of iterations is a nice value to at least be defined as a const at the top or something.

Reply:
At least for me, putting constants for magic numbers like this as a constant the top of the file makes it easier to see what's configurable. A comment on how 3 was chosen might also be useful.

Reply:
The comment is in line 117 below, which if not helpful if you are reading line 113, of course. I'll move the comment.